### ISSUE TITLE
Diff between Spring Cloud Config 1.0.6 and 1.1.0 when embedded in another application 

### ISSUE THREAD

**Issue by lquerel** [spring-cloud/spring-cloud-config#417]
Hi,

It seems that the last version of the Spring Cloud Config Server (1.1.0) prevents the configuration of custom REST controllers when embedded in another application.

My setup with the Cloud Config Server 1.0.6:
- SpringBoot app based on version 1.3.0 of springboot jars.
- Config Server embedded in this app with the annotation @EnableConfigServer
- Config Server prefix configured as follow: spring.cloud.config.server.prefix: /config
- Two REST controllers implemented by the app itself to expose some custom endpoint without collision with the prefix /config.

This setup works perfectly with the version 1.0.6 of Spring Cloud Config.

The same setup with Spring Cloud Config Server 1.1.0 and SpringBoot 1.3.5 doesn't work as I expected. The config server exposes all the endpoints under /config but my custom controllers are no longer exposed. 

When I remove the annotation @EnableConfigServer my custom controllers reappears but obviously the endpoints of the configuration server disappear. 

Is it still possible to embed the cloud config server in another SpringBoot App or this feature has been remove?

Thanks

Laurent Querel

**Comment by lquerel** [comment]
In addition to my previous message.

With this set of annotations the controller MyResourceController is found by Spring.

@SpringBootApplication(scanBasePackageClasses = {MyConfigurationServer.class, MyResourceController.class})
@EnableAutoConfiguration
//@EnableConfigServer

With the annotation @EnableConfigServer uncommented then the same controller is no longer found.

**Comment by spencergibb** [comment]
@lquerel I just tried it with 1.1.0 (Brixton.RELEASE) and 1.1.1 (Brixton.SR1) and I can add controllers that are outside of `/config`. Can you provide a sample that reproduces the problem?

**Comment by lquerel** [comment]
Made some progress here. My interpretation of the issue was incorrect but clearly I identified an issue when the @EnableConfigServer annotation is applied to my SpringBoot app. I attached a small example reproducing the issue.

My app contains two classes: MyApp and service named "ResourceRepository"... (Yes I know it's common name for a class but the java package hierarchy is there to avoid name collision :). I

This App works well when started without the annotation @EnableConfigServer 
But when the annotation is attached to MyApp then I get this exception.

org.springframework.beans.factory.UnsatisfiedDependencyException: Error creating bean with name 'resourceRepository' defined in class org.springframework.cloud.config.server.config.ResourceRepositoryConfiguration: Unsatisfied dependency expressed through constructor argument with index 0 of type [org.springframework.cloud.config.server.environment.SearchPathLocator]: No qualifying bean of type [org.springframework.cloud.config.server.environment.SearchPathLocator] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}; nested exception is org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type [org.springframework.cloud.config.server.environment.SearchPathLocator] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:749) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:464) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1123) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1018) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:510) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:482) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory$1.getObject(AbstractBeanFactory.java:306) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:230) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:302) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:197) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:772) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:839) ~[spring-context-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:538) ~[spring-context-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.boot.context.embedded.EmbeddedWebApplicationContext.refresh(EmbeddedWebApplicationContext.java:118) ~[spring-boot-1.3.5.RELEASE.jar:1.3.5.RELEASE]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:766) [spring-boot-1.3.5.RELEASE.jar:1.3.5.RELEASE]
	at org.springframework.boot.SpringApplication.createAndRefreshContext(SpringApplication.java:361) [spring-boot-1.3.5.RELEASE.jar:1.3.5.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:307) [spring-boot-1.3.5.RELEASE.jar:1.3.5.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1191) [spring-boot-1.3.5.RELEASE.jar:1.3.5.RELEASE]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1180) [spring-boot-1.3.5.RELEASE.jar:1.3.5.RELEASE]
	at test.MyApp.main(MyApp.java:16) [main/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_25]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_25]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_25]
	at java.lang.reflect.Method.invoke(Method.java:483) ~[na:1.8.0_25]
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144) [idea_rt.jar:na]
Caused by: org.springframework.beans.factory.NoSuchBeanDefinitionException: No qualifying bean of type [org.springframework.cloud.config.server.environment.SearchPathLocator] found for dependency: expected at least 1 bean which qualifies as autowire candidate for this dependency. Dependency annotations: {}
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.raiseNoSuchBeanDefinitionException(DefaultListableBeanFactory.java:1373) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.doResolveDependency(DefaultListableBeanFactory.java:1119) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.resolveDependency(DefaultListableBeanFactory.java:1014) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.resolveAutowiredArgument(ConstructorResolver.java:813) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	at org.springframework.beans.factory.support.ConstructorResolver.createArgumentArray(ConstructorResolver.java:741) ~[spring-beans-4.2.6.RELEASE.jar:4.2.6.RELEASE]
	... 24 common frames omitted


If I rename my class ResourceRepository to ResourceRepository2 then this issue disappear. Same thing if I expose this class an explicit name for the service then the issue disappear.

@Service("repo2")
class ResourceRepository

In conclusion when @EnableConfigServer is used you can't name a class ResourceRepository in your app independently of the package hierarchy. This issue doesn't exist with a classic SpringBoot App. 

Hope that help to fix this issue.

Thanks

Laurent 
 
[Archive.zip](https://github.com/spring-cloud/spring-cloud-config/files/316807/Archive.zip)

**Comment by spencergibb** [comment]
You happened on the name of a bean in Config Server. I don't know that there is anything we can do on our side to fix this. @dsyer?

**Comment by lquerel** [comment]
I renamed my class differently as a workaround. But I'm pretty sure that this kind of issue can occur for many other users with this class name or with another one.

May be used a fully qualified name instead of the name of the class could solve this issue.

**Comment by spencergibb** [comment]
Spring Cloud and Spring Boot create dozens of beans with all sorts of bean names. I don't think that renaming one bean name is a viable solution.

**Comment by spencergibb** [comment]
[beans.txt](https://github.com/spring-cloud/spring-cloud-config/files/316898/beans.txt)

**Comment by lquerel** [comment]
Don't understand why some beans are fully qualified and some others are not (see your beans.txt). I always thought that all beans were fully qualified by default. If it's not the case then I need to check that the names of my services, repositories, beans... (idependently of the package hierarchy) don't clash with the name of beans exposed by Spring Cloud and Spring Boot for the current version and for each future versions.

**Comment by spencergibb** [comment]
The beans that aren't fully qualified are beans created in an `@Configuration`. Many of the fully qualified end in `Configuration`, so they are configuration classes.
